File: 251003/Khlyshchankou/src/main/java/com/bsuir/dc/controller/PostController.java
package com.bsuir.dc.controller;

import com.bsuir.dc.dto.request.PostRequestTo;
import com.bsuir.dc.dto.response.PostResponseTo;
import com.bsuir.dc.service.PostService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/v1.0/posts")
public class PostController {
    private final PostService postService;

    @Autowired
    public PostController(PostService postService) { this.postService = postService; }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public PostResponseTo createPost(@RequestBody @Valid PostRequestTo postRequestTo) { return postService.save(postRequestTo); }

    @GetMapping
    @ResponseStatus(HttpStatus.OK)
    public List<PostResponseTo> getAllPosts() { return postService.findAll(); }

    @GetMapping("/{id}")
    @ResponseStatus(HttpStatus.OK)
    public PostResponseTo getPostById(@PathVariable Long id) { return postService.findById(id); }

    @PutMapping
    @ResponseStatus(HttpStatus.OK)
    public PostResponseTo updatePost(@RequestBody @Valid PostRequestTo postRequestTo){ return postService.update(postRequestTo); }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deletePost(@PathVariable long id){ postService.deleteById(id); }
}

File: 251003/Pankratiev/rest/discussion/src/main/kotlin/bsuir/dc/discussion/controller/PostController.kt
package bsuir.dc.discussion.controller

import bsuir.dc.discussion.dto.from.PostRequestTo
import bsuir.dc.discussion.dto.to.PostResponseTo
import bsuir.dc.discussion.service.PostService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import jakarta.validation.Valid

@RestController
@RequestMapping("/api/v1.0/posts")
class PostController(
    private val postService: PostService
) {
    @PostMapping
    fun createPost(
        @RequestBody @Valid postRequestTo: PostRequestTo,
        @RequestHeader("X-Country", defaultValue = "Belarus") country: String
    ): ResponseEntity<PostResponseTo> {
        val createdPost = postService.createPost(postRequestTo, country)
        println(createdPost)
        return ResponseEntity(createdPost, HttpStatus.CREATED)
    }

    @GetMapping
    fun getAllPosts(): ResponseEntity<List<PostResponseTo>> {
        val posts = postService.getAllPosts()
        return ResponseEntity.ok(posts)
    }

    @GetMapping("/{id}")
    fun getPostById(
        @PathVariable id: Long,
        @RequestHeader("X-Country", defaultValue = "Belarus") country: String
    ): ResponseEntity<PostResponseTo> {
        val post = postService.getPostById(country, id)
        return ResponseEntity.ok(post)
    }

    @DeleteMapping("/{id}")
    fun deletePost(
        @PathVariable id: Long,
        @RequestHeader("X-Country", defaultValue = "Belarus") country: String
    ): ResponseEntity<Void> {
        postService.deletePost(country, id)
        return ResponseEntity.noContent().build()
    }

    @PutMapping()
    fun updatePost(
        @RequestBody @Valid postRequestTo: PostRequestTo,
        @RequestHeader("X-Country", defaultValue = "Belarus") country: String
    ): ResponseEntity<PostResponseTo> {
        val updatedPost = postService.updatePost(country, postRequestTo)
        return ResponseEntity.ok(updatedPost)
    }
}
